core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -769,6 +800,75 @@ private SubPlan optimizePlan(SubPlan plan)
         {
             // Re-optimize plan here based on available runtime statistics.
             // Fragments changed due to re-optimization as well as their downstream stages are expected to be assigned new fragment ids.
+            plan = updateStagesPartitioning(plan);
+            return plan;
+        }
+
+        private SubPlan updateStagesPartitioning(SubPlan plan)
+        {
+            if (!runtimeAdaptivePartitioningEnabled || runtimeAdaptivePartitioningApplied) {
+                return plan;
+            }
+
+            for (SubPlan subPlan : planInTopologicalOrder) {
+                PlanFragment fragment = subPlan.getFragment();
+                if (!consumesHashPartitionedInput(fragment)) {
+                    // no input hash partitioning present
+                    continue;
+                }
+
+                StageId stageId = getStageId(fragment.getId());
+                if (stageExecutions.containsKey(stageId)) {
+                    // already started
+                    continue;
+                }
+
+                IsReadyForExecutionResult isReadyForExecutionResult = isReadyForExecution(subPlan);
+                // Caching is not only needed to avoid duplicate calls, but also to avoid the case that a stage that
+                // is not ready now but becomes ready when updateStageExecutions.
+                // We want to avoid starting an execution without considering changing the number of partitions.
+                // TODO: think about how to eliminate the cache
+                isReadyForExecutionCache.put(subPlan, isReadyForExecutionResult);
```

Comment:
`isReadyForExecutionCache.computeIfAbsent(...)` ?

Reply:
We probably can put the logic of how the cache is used into `isReadyForExecution` itself and just use the method. And clear the cache on each scheduler call.

Reply:
> isReadyForExecutionCache.computeIfAbsent(...) ?

The cache is completely fresh here.

Reply:
> isReadyForExecutionCache.computeIfAbsent(...) ?

Cache is fresh at the beginning.